### PR TITLE
Supports most recent Formatting API

### PR DESCRIPTION
**Pull request by konn**
Recent VSCode provides different APIs for ranges and whole-file formatting.
This pull-request overcomes that change and switched to use stdin/out to avoid unnecessary saves.

### CODE REVIEW

**Comment by mnxn** [comment]
Hi @konn,

Thanks for trying to improve this extension, it's much appreciated.

There are a couple of problems that need to be fixed before It can be merged.

1. Formatting selections doesn't seem to work at all

2. If I open a .hs file by itself, in other words not a workspace folder, the extensions produces a stack trace in the developer console and doesn't format.

File: src/formatter.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      stdin = input;
+      args.push("--write-mode", "display");
+    } else {
+      args.push(input.uri.fsPath);
+      args.push("--write-mode", "display");
+      maybeWorkspaceFolder = workspace.getWorkspaceFolder(input.uri);
+    }
+
+    const dir: string =
+      maybeWorkspaceFolder !== undefined
```
Comment (mnxn) [inline src/formatter.ts]:
```suggestion
      maybeWorkspaceFolder !== null
```

Should this be `null`?
When `maybeWorkspaceFolder` is declared, it uses a value of `null`. I believe `undefined` is only for uninitialized variables in js, so `maybeWorkspaceFolder` will never be `undefined`

Comment (konn) [reply src/formatter.ts]: Ah, that's right!

**Comment by konn** [comment]
Sorry for the late reaction. I think I fixed the behaviour of range-formatting and it seems that odd behaviour on a single file.

**Comment by mnxn** [comment]
Looks like it works now.

Merged.

I'll release to vscode in a minute

**Pull request merged**